modules/dnn/src/layers/layers_common.simd.hpp

```diff
@@ -737,5 +737,554 @@ void fastGEMM( const float* aptr, size_t astep, const float* bptr,
+        temp6 = vfredsum_vs_f32m2_f32m1(temp6, vs6, zero, vl);
+        temp7 = vfredsum_vs_f32m2_f32m1(temp7, vs7, zero, vl);
+
+        float32_t sum[8];
+        sum[0] = vfmv_f_s_f32m1_f32(temp0);
+        sum[1] = vfmv_f_s_f32m1_f32(temp1);
+        sum[2] = vfmv_f_s_f32m1_f32(temp2);
+        sum[3] = vfmv_f_s_f32m1_f32(temp3);
+        sum[4] = vfmv_f_s_f32m1_f32(temp4);
+        sum[5] = vfmv_f_s_f32m1_f32(temp5);
+        sum[6] = vfmv_f_s_f32m1_f32(temp6);
+        sum[7] = vfmv_f_s_f32m1_f32(temp7);
+
+        vfloat32m2_t s0 = vfadd_vv_f32m2(vle32_v_f32m2(sum, mvl), vle32_v_f32m2(bias + i, mvl), mvl);
+        vse32_v_f32m2(dst + i, s0, mvl);
+    }
+}
+
+enum { FASCONV_BASE_VECSZ = 4 }; // TODO: Large base size.
+void fastConv( const float* weights, size_t wstep, const float* bias,
+               const float* rowbuf, float* output, const int* outShape,
+               int blockSize, int vecsize, int vecsize_aligned,
+               const float* relu, bool initOutput )
+{
+    int vl = 4;
+    int outCn = outShape[1];
+    size_t outPlaneSize = outShape[2]*outShape[3];
+    float r0 = 1.f, r1 = 1.f, r2 = 1.f;
+    vfloat32m1_t vr0 = vfmv_v_f_f32m1(1, vl), vr1 = vfmv_v_f_f32m1(1, vl), vr2 = vfmv_v_f_f32m1(1, vl);
+    int maskbuf[FASCONV_BASE_VECSZ] = {0};
+    int rsz = blockSize % FASCONV_BASE_VECSZ;
+    for( int i = 0; i < rsz; i++ )
+        maskbuf[FASCONV_BASE_VECSZ - i - 1] = -1;
+    vint32m1_t vmaskbuf = vle32_v_i32m1(maskbuf ,vl);
+    vbool32_t mask = vmslt_vx_i32m1_b32(vmaskbuf, 0, vl); // mask for tail
+    // now compute dot product of the weights
+    // and im2row-transformed part of the tensor
+    for( int i = 0; i < outCn; i += 3 )
+    {
+        const float* wptr0 = weights + i*wstep;
+        const float* wptr1 = wptr0 + wstep;
+        const float* wptr2 = wptr1 + wstep;
+        float* outptr0 = output + i*outPlaneSize;
+        float* outptr1 = outptr0 + outPlaneSize;
+        float* outptr2 = outptr1 + outPlaneSize;
+        float bias0 = bias[i], bias1 = bias[i+1], bias2 = bias[i+2];
+
+        if( i+2 >= outCn )
+        {
+            wptr2 = wptr1;
+            outptr2 = outptr1;
+            bias2 = bias1;
+            if( i+1 >= outCn )
+            {
+                wptr2 = wptr1 = wptr0;
+                outptr2 = outptr1 = outptr0;
+                bias2 = bias1 = bias0;
+            }
+        }
+
+        if( relu )
+        {
+            r0 = relu[i]; r1 = relu[i+1]; r2 = relu[i+2];
+            if( i+2 >= outCn )
+            {
+                r2 = r1;
+                if( i+1 >= outCn )
+                    r2 = r1 = r0;
+            }
+            vr0 = vfmv_v_f_f32m1(r0, vl);
+            vr1 = vfmv_v_f_f32m1(r1, vl);
+            vr2 = vfmv_v_f_f32m1(r2, vl);
+        }
+
+        int j = 0;
+        for( ; j < blockSize; j += FASCONV_BASE_VECSZ )
+        {
+            bool tail = false;
+            if (j + FASCONV_BASE_VECSZ > blockSize)
+            {
+                if (j == 0) {
```

Comment:
Why do you need this branch? It case if `blockSize` is too small, tail should be called without loop iteration.

Reply:
I would like to explain why there is a branch about `j==0`.

When `j + 4 > blockSize`, it usually means we meet a tail then we use a tail mask and let `j = blockSize - 4` (compute the last 4 elements and store with mask). 

Example:

- Assume blockSize=5 and output is 5*1, ignore the output channel
- The mask will be `[0, 0, 0, 1]`
- The first loop iteration will compute the first 4 elements in the output matrix, then the output should be `[√, √, √, √, TBD]`. Now, j = 4.
- The second also the last loop iteration will compute the last element with the help of the mask. Actually, we compute the last 4 elements but only store the last one because of the mask. In detail, at the beginning of this iteration, j=4 and we find that `j + 4 > blockSize`, so we let `j = blockSize - 4`, now, j=1 then we compute the last 4 elements and only store the last one.

However, there is another situation here when the blockSize is smaller than 4. In this case, we should not use mask but use `vl`. I will explain with an example.

Example:

- Assume `blockSize=1 `and output is 1*1, also ignore the output channel.
- The mask will be also `[0, 0, 0, 1],` which can NOT use. 
- Instead, we let vl=1 and load-compute-store directly with the help of `vl`.